design/decisions/0005-entity-registration-api.md

```diff
@@ -0,0 +1,398 @@
+But, in environments where that is not feasible, the current model without feedback can have unpleasant side effects.
+
+For example, if an entity registers itself with topic id: `te/device/child01//` and `@id`: `child01`,
+a device twin with external id `child01` is created in the cloud and all of its data is routed to this twin.
+If another device accidentally registers itself using the same topic id, but with `@id`: `child02`,
+a second device twin is created in the cloud with the external id: `child02`
+and all the messages published to that topic by `child02` and even `child01` are routed to the new device twin.
+The `child01` twin would remain orphaned in the cloud, without the user noticing anything immediately, for lack of feedback.
+
+To avoid such issues, the registering entities must be made aware of any conflicts/errors immediately
+so that they don't proceed publishing any data to the wrong entity assuming that it succeeded.
+In this case, %%te&& must have rejected the second registration message.
+Any updates to an existing entity must have been done using dedicated entity update APIs.
+
+## Challenges
+
+* In the current architecture, the mapper handles all the registration messages and maintains the entity store.
+  So, in a deployment with multiple mappers, there would be multiple entity stores maintained by each mapper.
+  This results in unnecessary duplication of data and makes it error-prone with multiple sources of truth.
+  A central component like `tedge-agent` would be a better fit to store and maintain the entity store.
+* The current entity store also maintains the external id of devices, which is a cloud-specific aspect.
+  When the entity store is moved to the `tedge-agent`, this aspect must be detached and left with the mappers,
+  without the mappers having to maintain a duplicate entity store.
+* When the `tedge-agent` handles the registration of entities, it can't consult with the mappers to make sure that
+  the entities that it accepted locally are valid in the cloud as well.
+  If the local registration succeeds, but the cloud registration fails, the user must get that feedback as well.
+* A mapper might get connected much later, after a lot of local registrations are already processed.
+  Even in such cases, the mapper should be able to fetch all the entities that were registered till then,
+  and replicate that in the cloud.
+  Even in that case, any failed registrations must be conveyed to the user asynchronously.
+
+## Solution Proposal
+
+### Proposal 1: REST APIs for registration
+
+* The `tedge-agent` maintains the entity store and provides REST APIs for registration, and querying.
+  The mappers and other components can query entity data from it when needed.
+* On every successful registration, `tedge-agent` publishes those registration messages to their respective topics,
+  so that other components likes mappers subscribed to those topics are notified of those new entities.
```

Comment:
Does the agent also updates the entity store when a registration message is published by another component?

I would prefer not (to force the clients use the HTTP API), but this breaks backward compatibility.

If the agent also reacts on registration updates over MQTT, one must ensure the agent doesn't react on its own updates.

Reply:
> Does the agent also updates the entity store when a registration message is published by another component?

We can take a stance that those using the HTTP APIs for registration must stick to it for all registrations and not combine both HTTP and MQTT based explicit registrations. I still expect many users continuing to use the auto-registration mechanism over MQTT. But, I'm assuming that a very small subset of users are currently using explicit registrations and we can ask them to migrate to the HTTP APIs instead.

Reply:
I don't think doing a breaking change will be acceptable here. Some plugins are already doing this (e.g. [tedge-container-plugin](https://github.com/thin-edge/tedge-container-plugin)). And I don't see a "nice" transition phase.

Reply:
> I don't think doing a breaking change will be acceptable here.

So, the mapper has to:
- handle registration messages received over MQTT as `PATCH` requests,
- publish registration state over MQTT only after successful updates introducing an actual change.

Reply:
Addressed by eb0eba88aa10427f4eb7ad70a0c04677202395b5